### PR TITLE
Remove the DDP DataLoader

### LINKED ISSUE #244: Remove the `DDPDataLoader`
Let's remove the `DDPDataLoader` and instead move its functionality directly into the training loop -- e.g.

```python3

for epoch in epochs:
  if isinstance(state.train_dataloader.sampler, DistributedSampler):
    state.train_dataloader.sampler.set_epoch(int(state.timer.epoch))
```

### PR DESCRIPTION
**Pull request by ravi-mosaicml**
Closes #244

### CODE REVIEW

**Review by jbloxham — APPROVED** [review]
Looks good - but I do wonder if down the road we could remove some of the repetition by repurposing the callbacks system in the trainer.

File: composer/trainer/trainer.py [hunk 1]
```diff
@@ -807,6 +806,11 @@ def eval(self, is_batch: bool):
 
             metrics = self._get_metrics_as_collection(is_train=False)
 
+            if isinstance(self.state.eval_dataloader.sampler, torch.utils.data.DistributedSampler):
+                # using the batch as the epoch, as the evaluator may be running every batch
+                # instead of every epoch
+                self.state.eval_dataloader.sampler.set_epoch(int(self.state.timer.batch))
+
```
Comment (abhi-mosaic) [inline composer/trainer/trainer.py:813]: What is this section of code meant for? The eval dataloader's sampler should probably not matter. And setting the epoch to the current batch seems a bit strange?

Comment (ravi-mosaicml) [reply composer/trainer/trainer.py]: Discussed offline; updated comment.

**Comment by ravi-mosaicml** [comment]
> Looks good - but I do wonder if down the road we could remove some of the repetition by repurposing the callbacks system in the trainer.

We discussed this a while back (I think before you started) and decided against this design so it becomes explicit in the training loop what's happening where.

**Pull request merged**